CMake build system

mzuzek opened the pull request:
Fixes #128

## Summary

With this PR we enable building the tools with CMake. Caliper and Apex are connected as git submodules and can be built from source if no existing installation is found. In build, tools try to find Kokkos and propagate it's settings (e.g. enable CUDA support) to tools and TPLs. "Single library intefrace" is introduced which allows simple linking of all enabled tools and TPLs to client application.

### Build: sample app CMake list file

```cmake
find_package(Kokkos REQUIRED)
add_directory(kokkos-tools) # TODO: CMake install exports

add_executable(myapp source.cpp)

target_link_libraries(myapp Kokkos::kokkos kokkostools) # link all tools
```

### Usage: sample app source code

```cpp
#include <Kokkos_Core.hpp>
#include "kp_all.hpp"

int main(int argc, char *argv[])
{
  const char *profiler_name = "caliper";
  const char *profiler_config = "runtime-report(profile.kokkos)";

  auto eventSet = KokkosTools::get_event_set(profiler_name, profiler_config);
  Kokkos::Tools::Experimental::set_callbacks(eventSet);

  Kokkos::initialize(argc, argv);
  // ... run app ... //
  Kokkos::finalize();
  return 0;
}
```

## Contents / Status

* General:
  - [x] CMake scripts to build tools;
  - [x] Single library interface linking all tools and TPLs;
  - [x] Fetch basic Kokkos settings in alternative scenarios:
    - [x] 1. from existing installation (`find_package(Kokkos)` based on `Kokkos_ROOT`);
    - [x] 2. from source if Kokkos is being built within same parent CMake project;
    - [x] 3. skip option propagation if Kokkos is not available;
  - [x] MPI support;
  - [x] Basic Windows (MSVC) build support - with Caliper and VTune;
  - [x] CMake exports for tools Installation;
  - [ ] Allow building each tool separately;
* Building tools (for traditional runtime shared library connection):
  - [x] Native tools: `space-time-stack`, `chrome-tracing`, `simple-kernel-timer`, `simple-kernel-timer-json`, `memory-hwm-mpi`, `memory-usage`, `memory-events`, `memory-hwm` and `memory-hwm-mpi`;
  - [x] Native utilities: `kernel-filter` and `kernel-logger`;
  - [x] TPL connectors:  [VTune](https://www.intel.com/content/www/us/en/develop/documentation/vtune-help/top.html), [SystemTap](https://sourceware.org/systemtap/), [Variorum](https://github.com/llnl/variorum) and [PAPI](https://bitbucket.org/icl/papi/wiki/Home);
  - [x] TPLs: [Caliper](https://github.com/LLNL/Caliper) and [Apex](https://github.com/UO-OACISS/apex) built from source (submodules) if no existing installation is detected;
  - [ ] TPLs: [apollo (`builtin-ml-library` branch)](https://github.com/llnl/apollo/tree/builtin-ml-library);
  - [ ] GPU profiler connectors:
    - [x] [_nvprof_](https://docs.nvidia.com/cuda/profiler-users-guide/index.html#nvprof-overview) (CUDA);
    - [ ] [_roctx_](https://rocmdocs.amd.com/en/latest/Installation_Guide/Tools-Installation.html#roctracer) (HIP);
* Tools available via `KokkosTools::get_event_set(profiler_name, profiler_config)` single library interface:
  - [x] All native tools;
  - [x] TPL connectors: VTune, SystemTap and Variorum;
  - [x] Caliper with configuration string support, requires WIP https://github.com/LLNL/Caliper/pull/402 branch;
  - [ ] Apex, PAPI connector and GPU connectors;
  - [ ] Native utilities (`kernel-filter` and `kernel-logger`) + implement `std::vector`/array overload of `Kokkos::Tools::Experimental::set_callbacks()` for registering multiple event sets ?
* Documentation and examples:
  - [x] Add [example](https://github.com/NexGenAnalytics/kokkos-tools/tree/cmake-build-system/example) demonstrating single library interface;
  - [ ] Update [ReadMe.md](https://github.com/kokkos/kokkos-tools/blob/master/README.md) with instructions on how to configure and build tools with CMake;
* Other / fixes:
  - [ ] Implement own tool to fetch parent path and replace `cmake_path(GET ... PARENT_PATH ...)` which requires CMake v3.20 ?

#### Future considerations:
- connect [_timemory_](https://timemory.readthedocs.io/en/develop/) (like _Caliper_ and _Apex_);
- consider introducing lazy profiler initialization (`kokkosp_init_library()` called after `Kokkos::initialize()`);

Commit:
CMake list files for basic profilers #128

Commit:
CMake list file for PAPI connector #128

Commit:
CMake list file for Variorum connector #128

dalg24 commented on CMakeLists.txt:
```diff
@@ -0,0 +1,74 @@
+
+# Set policies
+cmake_policy(SET CMP0054 NEW) # simplify if()
+cmake_policy(SET CMP0057 NEW) # support IN_LIST in if()
```
These two are pointless if we require cmake 3.16

vlkale replied:
Removed.

vlkale replied:
Removed in PR.

Commit:
PAPI find module fixes #128

Commit:
Fixed module path #128

Commit:
Fixed VTune_ROOT #128

Commit:
Caliper submodule and configuration #128

Commit:
Find MPI, OpenMP and PAPI in main CMake file #128

Commit:
Apex submodule and configuration #128

Commit:
Tools and Kokkos options (Kokkos in parent project) #128

DavidPoliakoff commented on CMakeLists.txt:
```diff
@@ -0,0 +1,169 @@
+if(TARGET Kokkos::kokkos)
+  message(STATUS "Using Kokkos settings for KokkosTools (source: ${Kokkos_SOURCE_DIR})")
+else()
+  message(WARNING "Not using Kokkos for tools configuration\n\tNote: using settings from installed is Kokkos not yet supported")
```
Oh, this is a really slick idea

jrmadsen replied:
Why would there be a `Kokkos::Kokkos` target w/o a `find_package(Kokkos)`?

DavidPoliakoff replied:
In my app's CMakeLists.txt

```
add_subdirectory(kokkos) # defines the target
add_subdirectory(kokkos-tools)
```

jrmadsen replied:
That's not very clear, especially with the `set(CMAKE_CXX_COMPILER ${Kokkos_CXX_COMPILER})` down below.

DavidPoliakoff replied:
Yeah, could definitely use a comment. It's also more of a Lawrence Livermore-ism than something Sandia codes do today

MikolajZuzek replied:
I cleaned this up a bit (excuse messy draft). The idea is that:
1. we check if Kokkos is built within same parent project - and before Tools, as in David's example above. Is there a better way than checking for `TARGET Kokkos::kokkos` ? In this case we're all set up with options, compiler and flags.
2. we check for installed Kokkos in `Kokkos_ROOT` and extract all relevant settings from there.
3. in case Kokkos is nowhere found, we might just built Tools alone (figuring reasonable option defaults) or error out.

Any suggestions appreciated (where's the slicky part ?)

DavidPoliakoff commented:
I like where you're starting with this. Also looping in @daboehme in case he has opinions about how we're playing with Caliper here

jrmadsen commented on CMakeLists.txt:
```diff
@@ -0,0 +1,169 @@
+message(STATUS)
+
+# Tools settings
+set(KOKKOSTOOLS_ENABLE_CALIPER ON CACHE BOOL "Enable building Caliper library")
+set(KOKKOSTOOLS_ENABLE_APEX    ON CACHE BOOL "Enable building Apex library")
```
```suggestion
option(KOKKOSTOOLS_ENABLE_CALIPER "Enable building Caliper library" ON)
option(KOKKOSTOOLS_ENABLE_APEX    "Enable building Apex library"    ON)
```

jrmadsen commented on CMakeLists.txt:
```diff
@@ -0,0 +1,169 @@
+# CXXFLAGS=-O3 -std=c++11 -g -shared -fPIC
+# -Wall -Wextra
+
+set(CMAKE_CXX_COMPILER ${Kokkos_CXX_COMPILER})
```
I _really_ wouldn't recommend doing this. Emit an `AUTHOR_WARNING` (so it can be suppressed) but don't force the same compiler bc you might end up with tools which have bugs w/ certain compilers/compiler versions. IIRC, the whole reason @DavidPoliakoff was so strict about KokkosTools having a C interface was avoid issues w/ KokkosTools being compiled with a different C++ compiler as kokkos.

MikolajZuzek replied:
Thanks. I started with the idea that we should strictly enforce same compiler and flags, then got course-corrected toward more flexible approach. I'm removing this for now, following your suggestion.

> Note: Instead of wiping or commenting that out I introduced `KokkosTools_REUSE_KOKKOS_COMPILER` advanced option as a way for the code to show our intentions (where the discussion stands).

jrmadsen commented on CMakeLists.txt:
```diff
@@ -0,0 +1,169 @@
+# -Wall -Wextra
+
+set(CMAKE_CXX_COMPILER ${Kokkos_CXX_COMPILER})
+set(CMAKE_CXX_STANDARD ${CMAKE_CXX_STANDARD_DEFAULT})
```
Cache this so it can be changed. You cannot override a local variable with a `-DCMAKE_CXX_STANDARD=...` on the command line.

MikolajZuzek replied:
Fixed, thank you.

jrmadsen commented on CMakeLists.txt:
```diff
@@ -0,0 +1,169 @@
+set(CMAKE_CXX_STANDARD ${CMAKE_CXX_STANDARD_DEFAULT})
+
+# Libraries
+find_package(PAPI)
```
There should be an option to enable/disable papi support

jrmadsen replied:
This should (IMO) take the form of:

```cmake
option(KOKKOSTOOLS_USE_PAPI "Enable PAPI support" ON)
if(KOKKOSTOOLS_USE_PAPI)
    find_package(PAPI REQUIRED)
endif()
```

So I used to do what you did here in the past. Experience has taught me that it is _much_ better to be explicit: make it an option and fail if it is not found. Especially since you have Caliper and Apex subprojects -- users will likely miss that PAPI was not found bc there will be so many messages from adding those subprojects.

MikolajZuzek replied:
Fixed, thank you (PAPI is now guarded with the option as suggested).

jrmadsen commented on CMakeLists.txt:
```diff
@@ -0,0 +1,169 @@
+  cmake_path(GET PAPI_INCLUDE_DIR PARENT_PATH PAPI_ROOT)
+endif()
+find_package(OpenMP)
+find_package(MPI)
```
There should be an option to enable/disable MPI support

MikolajZuzek replied:
Right, there's `USE_MPI` in source files. I'm working on that and will report back here.

MikolajZuzek replied:
Added `KokkosTools_ENABLE_MPI` option in deb807bd501adc8626736d9f2e9ff700c10c6ad8 and based C++ `USE_MPI` on that (through config header).

jrmadsen commented on CMakeLists.txt:
```diff
@@ -0,0 +1,169 @@
+if(PAPI_FOUND)
+  cmake_path(GET PAPI_INCLUDE_DIR PARENT_PATH PAPI_ROOT)
+endif()
+find_package(OpenMP)
```
I'm not sure where OpenMP is needed but there should be an option to enable/disable OpenMP support.

MikolajZuzek replied:
OpenMP is not actually used by tools, although it's mentioned in [`memory-usage` Makefile](https://github.com/kokkos/kokkos-tools/blob/7dc3f1996b0928fc8ac440c4d84cf62774dcbe35/profiling/memory-usage/Makefile#L2)
Cleaned that up on 713ad680d32d52327e034c25201fe13fe0df5d52

jrmadsen commented on CMakeLists.txt:
```diff
@@ -0,0 +1,169 @@
+  #set(BUILD_DOCS OFF)   # Build documentation.
+  #set(BUILD_TESTING OFF) # Build unit tests.
+  #set(WITH_FORTRAN OFF) # Build and install Fortran wrappers.
+  set(WITH_KOKKOS ON) # Enable Kokkos profiling support
```
So I added support in Caliper to do:

```cmake
set(CALIPER_OPTION_PREFIX ON CACHE INTERNAL "Prefix caliper options with CALIPER_" FORCE)
```

and all the caliper options, e.g. `WITH_KOKKOS` appear in the cache as `CALIPER_WITH_KOKKOS`. This makes it very clear what are caliper options when it is a subproject.

MikolajZuzek replied:
Great, thanks! Options prefixed.

jrmadsen commented on CMakeLists.txt:
```diff
@@ -0,0 +1,169 @@
+  set(WITH_KOKKOS ON) # Enable Kokkos profiling support
+
+  # Build with support for the OpenMP tools interface.
+  if(OpenMP_FOUND)
+    set(WITH_OMPT ON)
+  else()
+    set(WITH_OMPT OFF)
+  endif()
```
```suggestion
```

Let the user specify `CALIPER_WITH_OMPT=ON`, there is no way to override this local variable if there is a problem with the Caliper OMPT build

MikolajZuzek replied:
Thanks, fixed on 713ad680d32d52327e034c25201fe13fe0df5d52

jrmadsen commented on CMakeLists.txt:
```diff
@@ -0,0 +1,169 @@
+    set(WITH_OMPT OFF)
+  endif()
+
+  set(WITH_SAMPLER ON)  # Enable time-based sampling on Linux.
+  set(WITH_TOOLS ON)    # Build Caliper’s tools (i.e, cali-query and mpi-caliquery). Default: On.
+  set(WITH_MPI ${KOKKOS_ENABLE_MPI}) # Build with MPI support.
+  set(WITH_ROCM ${Kokkos_ENABLE_HIP}) # Enable AMD ROCtracer/RocTX support
+  set(WITH_NVTX ${Kokkos_ENABLE_CUDA}) # Build adapters to forward Caliper annotations to NVidia’s nvtx annotation API. Set CUDA_TOOLKIT_ROOT_DIR to the CUDA installation.
+  set(WITH_CUPTI ${Kokkos_ENABLE_CUDA}) # Enable support for CUDA performance analysis
+                                        # (wrapping of driver/runtime API calls and CUDA activity tracing).
+  # Enable PAPI hardware counter service (requires papi)
+  if(PAPI_FOUND)
+    set(WITH_PAPI ON)
+    set(PAPI_PREFIX ${PAPI_ROOT})
+  endif()
+  # Build adapters to forward Caliper annotations to Intel’s VTune annotation API.
+  if(VTune_ROOT)
+    set(WITH_VTUNE ON)
+    set(ITT_PREFIX ${VTune_ROOT}/sdk)
+    # TODO: Caliper doesn't seek ITT in ${VTune_ROOT}/sdk/lib64 or ${VTune_ROOT}/sdk/lib32 (only in .../lib)
+  else()
+    set(WITH_VTUNE OFF)
+  endif()
```
Again, you are forcing a configuration which the user may not want. Let them specify it themselves or, at the very least, set these as cache variables so there are default settings but the user can change them.

MikolajZuzek replied:
I've changed all settings to cache variables and also minimized number of option defaults imposed on Caliper and Apex by the tools: it nails down to enabling support for Kokkos, MPI, Variorum, VTune, PAPI and CUDA/HIP support - when those are requested/detected (also the compiler if user enables `KokkosTools_REUSE_KOKKOS_COMPILER`).

jrmadsen commented on CMakeLists.txt:
```diff
@@ -0,0 +1,169 @@
+# Note: Apex docs http://uo-oaciss.github.io/apex/install/#standalone_installation
+macro(configure_apex)
+  #set(BUILD_STATIC_EXECUTABLES OFF)
+  set(APEX_WITH_PAPI ${PAPI_FOUND})
+  set(APEX_WITH_CUDA ${Kokkos_ENABLE_CUDA})
+  set(APEX_WITH_BFD ON)
+  #set(BFD_ROOT ...)
+  if(OMPT_FOUND)
+    set(APEX_WITH_OMPT ON)
+    # TODO: set(OMPT_ROOT ...) ?
+  else()
+    set(APEX_WITH_OMPT OFF)
+  endif()
+  # TODO: Apex tries to fetch TPLs and install them in /usr/lib
+  #       (normally failing) unless we disable them...
+  set(APEX_WITH_ACTIVEHARMONY OFF)
+  set(APEX_WITH_OTF2 OFF)
+  set(APEX_WITH_MPI OFF)
```
Same as before, cache variables so the user can configure them as needed.

jrmadsen commented on CMakeLists.txt:
```diff
@@ -0,0 +1,169 @@
+
+# Build TPLs
+if(KOKKOSTOOLS_ENABLE_CALIPER)
+  configure_caliper()
```
I don't really see the need for a macro here. Personally, i think you should just put setting the initial cache values in this if block.

MikolajZuzek replied:
My motivation behind those macros was to move Caliper and Apex setup to the other file and keep scripts tight (applied in 4b7a3441904710e7c6e50952bc265eca83380b97) - let me know if you don't like that approach.

jrmadsen commented on CMakeLists.txt:
```diff
@@ -0,0 +1,169 @@
+  add_subdirectory(tpls/Caliper)
+endif()
+if(KOKKOSTOOLS_ENABLE_APEX)
+  configure_apex()
```
Same comment about using a macro here

MikolajZuzek replied:
(see macro discussion above)

jrmadsen commented on cmake/FindPAPI.cmake:
```diff
@@ -0,0 +1,101 @@
+#]=======================================================================]
+
+# Look for the header file.
+if(NOT PAPI_INCLUDE_DIR)
```
```suggestion
```

You don't need to protect this.

jrmadsen commented on cmake/FindPAPI.cmake:
```diff
@@ -0,0 +1,101 @@
+  return()
+endif()
+
+# Skip target if already defined
+if(TARGET PAPI::libpapi)
+  return()
+endif()
+
+# Set up imported target
+add_library(PAPI::libpapi UNKNOWN IMPORTED)
+
+set_property(
+  TARGET PAPI::libpapi APPEND PROPERTY
+  IMPORTED_CONFIGURATIONS RELEASE DEBUG)
+
+set_target_properties(
+  PAPI::libpapi PROPERTIES
+  INTERFACE_INCLUDE_DIRECTORIES "${PAPI_INCLUDE_DIR}"
+  IMPORTED_LINK_INTERFACE_LANGUAGES "C"
+  IMPORTED_LOCATION "${PAPI_LIBRARY}"
+  IMPORTED_LOCATION_RELEASE "${PAPI_LIBRARY}"
+  IMPORTED_LOCATION_DEBUG "${PAPI_LIBRARY}")
```
```suggestion
# Set up imported target
add_library(PAPI::PAPI INTERFACE IMPORTED)

target_include_directories(PAPI::PAPI INTERFACE ${PAPI_INCLUDE_DIR})
target_link_libraries(PAPI::PAPI INTERFACE ${PAPI_LIBRARY})

set(PAPI_INCLUDE_DIRS ${PAPI_INCLUDE_DIR})
set(PAPI_LIBRARIES ${PAPI_LIBRARY})
```

This is how I typically do this.

MikolajZuzek replied:
Suggestion applied, thank you.

I'll leave `if(TARGET PAPI::PAPI)` guard there, because we'd get target collisions in Apex or Caliper otherwise: they find PAPI using our find module now - since we prioritized it on `CMAKE_MODULE_PATH` (see the other applied remark) - so it can be called multiple times.

jrmadsen commented on cmake/FindPAPI.cmake:
```diff
@@ -0,0 +1,101 @@
+  string(REPLACE "," "."  PAPI_VERSION_STRING "${PAPI_VERSION_STRING}")
+endif()
+
+find_package_handle_standard_args(
+  PAPI
+  REQUIRED_VARS PAPI_LIBRARY PAPI_INCLUDE_DIR
+  VERSION_VAR PAPI_VERSION_STRING
+  HANDLE_COMPONENTS)
```
```suggestion
include(FindPackageHandleStandardArgs)
find_package_handle_standard_args(
  PAPI
  REQUIRED_VARS PAPI_LIBRARY PAPI_INCLUDE_DIR
  VERSION_VAR PAPI_VERSION_STRING)
```

You do not support components so `HANDLE_COMPONENTS` is unnecessary.

jrmadsen commented on cmake/FindPAPI.cmake:
```diff
@@ -0,0 +1,101 @@
+  VERSION_VAR PAPI_VERSION_STRING
+  HANDLE_COMPONENTS)
+
+# Handle missing library
+if(NOT PAPI_FOUND)
+  if(PAPI_FIND_REQUIRED)
+    message(FATAL_ERROR "PAPI library not found")
+  elseif(NOT PAPI_FIND_QUIETLY)
+    message(WARNING "PAPI library not found")
+  endif()
+  return()
+endif()
```
```suggestion
```

This is handled by `find_package_handle_standard_args`

jrmadsen commented on cmake/FindPAPI.cmake:
```diff
@@ -0,0 +1,101 @@
+    NAMES papi.h
+    HINTS /usr/include /usr/local/include)
+  mark_as_advanced(PAPI_INCLUDE_DIR)
+endif()
```
```suggestion
```

MikolajZuzek replied:
Suggestion applied in b20828a09913424257c98cf33491de3199feb4f2

jrmadsen commented on cmake/FindPAPI.cmake:
```diff
@@ -0,0 +1,101 @@
+  mark_as_advanced(PAPI_INCLUDE_DIR)
+endif()
+
+if(NOT PAPI_LIBRARY)
```
```suggestion
```

MikolajZuzek replied:
Suggestion applied in b20828a09913424257c98cf33491de3199feb4f2

jrmadsen commented on cmake/FindPAPI.cmake:
```diff
@@ -0,0 +1,101 @@
+    NAMES papi
+    HINTS /usr/lib /usr/local/lib)
+  mark_as_advanced(PAPI_LIBRARY)
+endif()
```
```suggestion
```

MikolajZuzek replied:
Suggestion applied in b20828a09913424257c98cf33491de3199feb4f2

jrmadsen commented on profiling/chrome-tracing/CMakeLists.txt:
```diff
@@ -0,0 +1,7 @@
+
+add_library(kp_chrome_tracing SHARED kp_chrome_tracing.cpp)
+
+if(NOT DEFINED USE_MPI OR USE_MPI)
```
```suggestion
if(KOKKOSTOOLS_USE_MPI AND TARGET MPI::MPI_CXX)
```

MikolajZuzek replied:
Thanks, applied this suggestion in deb807bd501adc8626736d9f2e9ff700c10c6ad8.
See MPI support discussion above.

jrmadsen reviewed (commented):
Just my thoughts

jrmadsen commented on CMakeLists.txt:
```diff
@@ -0,0 +1,169 @@
+  message(FATAL_ERROR "FATAL: In-source builds are not allowed. You should create a separate directory for build files.")
+endif()
+
+list(APPEND CMAKE_MODULE_PATH ${CMAKE_CURRENT_SOURCE_DIR}/cmake)
```
```suggestion
list(INSERT CMAKE_MODULE_PATH 0 ${PROJECT_SOURCE_DIR}/cmake)
```

You should ensure that your `FindPAPI.cmake` is always picked up first.

Commit:
Use `option` to define CMake options #128

Co-authored-by: Jonathan R. Madsen <jrmadsen@users.noreply.github.com>

Commit:
Fix case in option names #128

Commit:
Option for PAPI support #128

Commit:
Fix forced compiler settings (allow user build options) #128

Commit:
Fix reduntant code in PAPI find module

covered in find_package_handle_standard_args()

Co-authored-by: Jonathan R. Madsen <jrmadsen@users.noreply.github.com>

Commit:
Fix case in option names #128

Commit:
Prefix Caliper options #128

Commit:
FIxed find modules pick up order #128

Co-authored-by: Jonathan R. Madsen <jrmadsen@users.noreply.github.com>

Commit:
Define interface target and clean up its properties #128

Co-authored-by: Jonathan R. Madsen <jrmadsen@users.noreply.github.com>

Commit:
Skip unnecesary guards in PAPI find module

Co-authored-by: Jonathan R. Madsen <jrmadsen@users.noreply.github.com>

Commit:
Skip [unused] components handling in PAPI find module

Co-authored-by: Jonathan R. Madsen <jrmadsen@users.noreply.github.com>

Commit:
Clean up Kokkos options extraction (still in progress) #128

Commit:
Move Caliper and Apex configuration to separate file #128

Commit:
MPI support enabled with KokkosTools_ENABLE_MPI #128

Commit:
Cleaned redundant find_package(PAPI) #128

Commit:
Fixed missing USE_MPI guards in Variorum connector #128

Commit:
Fix OpenMP Tools build errors: disable OMPT in Apex and leave default in Caliper

Commit:
MPI config fixes

Commit:
Add shared libs option

Commit:
Initial single library interface + connect Caliper

crtrott committed:
Merge pull request #133 from kokkos/develop

Develop

Commit:
Connect kernel-timer to single-lib interface

Commit:
Connect memory-usage to single-lib interface

Commit:
Adapt simple-kernel-timer for single library interface

Commit:
Rename activate_tool() to get_event_set()

Commit:
Connect memory-events to single-lib interface

Commit:
Connect HighwaterMark to single-lib interface

Commit:
Refactoring of native handlers in single-lib interface

Commit:
Connect chrome-tracing

Commit:
Connect space-time-stack

Commit:
Connect variorum-connector

Commit:
Connect systemtap-connector

Commit:
Move Variorum config script to cmake folder

Commit:
Fix missing MPI guard

Commit:
Connect vtune-connector and vtune-focused-connector to single-lib interface

Commit:
Build static libraries with BUILD_SHARED_LIBS=OFF

Commit:
Enable building on Windows

Commit:
Detect and use Kokkos installation, if available

Commit:
Find Caliper installations before building it

Commit:
Find existing Apex installations before building it

Commit:
Rename ITT to avoid name conflict with Apex

Commit:
Remove memory-usage dependency on kp_memory_events.hpp

Commit:
Configure Apex based on Kokkos settings

Commit:
Configure Caliper based on Kokkos settings

Commit:
Add sample application

Commit:
Update submodules

Commit:
Adjust for Caliper methods rename

Commit:
Fix nested namespaces (don't require c++17)

MikolajZuzek commented:
Thank you @jrmadsen for a thorough review and great comments!

MikolajZuzek commented:
@crtrott @jrmadsen @daboehme @khuck @DavidPoliakoff @fnrizzi 

At this point I was able to draft a CMake build system for KokkosTools with:
- shared libraries built for most native tools and connectors;
- single library interface that connects all native tools, Caliper and some TPL connectors;
- Kokkos settings acquired from existing installation or Kokkos included in parent project;
- sample app that links tools and TPLs with new single library interface (see [`example`](https://github.com/NexGenAnalytics/kokkos-tools/tree/cmake-build-system/example) folder) + CMake tests running this demo with all implemented tools;
- Windows/MSVC build tested with VTune and Caliper;

You'll find more details in [the PR description](https://github.com/kokkos/kokkos-tools/pull/131#issue-1040253019) (just cleaned it up). I hope this shows overall idea about how things are connected, configured and built: any suggestions and ideas you may have will be much appreciated. My default plan is to continue with connecting the remaining tools same way.

crtrott commented:
FYI: I think this looks really good. I think priority should be the NVIDIA connector tool and the CMake export.

crtrott committed:
Merge pull request #135 from kokkos/develop

Just update some things.

Commit:
Connect nvprof-connector

Commit:
Add CMake installation/exports

Commit:
Fix replicated BUILD_SHARED_LIBS effects

mzuzek commented:
@crtrott @dalg24 @fnrizzi 

Looks like we have the tasks listed in #128 covered to a decent extent.
Are there any immediate items you would see added or fixed on this PR ?

vlkale committed:
Fixed defaults for clean build.

vlkale committed:
Fix to newline at end of file

fnrizzi committed:
Merge pull request #2 from vlkale/cmake-build-system

dalg24 commented on CMakeLists.txt:
```diff
@@ -0,0 +1,220 @@
+message(STATUS Configuring Kokkos-Tools)
+message(STATUS)
+
+# Common settings
+option(BUILD_SHARED_LIBS "Build shared libraries" ON)
+if(WIN32)
+  set(BUILD_SHARED_LIBS OFF) # We need to add __declspec(dllexport/dllimport) for Windows DLLs
+endif()
```
What is the intent here?  `BUILD_SHARED_LIBS` is a global flag recognized by CMake.  It does not need to be an option.  Was that an attempt to change the default value to `ON`.
Also what is is the deal with Windows?  What would be the point to build static?

crtrott replied:
Regarding the option: yeah I think the idea was to have the default be ON.
And on Windows you can still link a static tools library into your app, and use the API to set callbacks. We may wanna actually set the KokkosTools_ENABLE_SINGLE default to ON for windows.

jrmadsen replied:
If you want to get semi-lazy for windows and avoid all the import/export mess, just set [CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS](https://cmake.org/cmake/help/latest/prop_tgt/WINDOWS_EXPORT_ALL_SYMBOLS.html) to ON.

vlkale replied:
> Regarding the option: yeah I think the idea was to have the default be ON. And on Windows you can still link a static tools library into your app, and use the API to set callbacks. We may wanna actually set the KokkosTools_ENABLE_SINGLE default to ON for windows.

Yes, that makes sense. I will set KokkosTools_ENABLE_SINGLE (or KokkosTools_ENABLE_MONOLITHIC?) to ON.

dalg24 commented on CMakeLists.txt:
```diff
@@ -0,0 +1,220 @@
+endif()
+
+# Tools settings
+option(KokkosTools_ENABLE_SINGLE  "Build single library interfacing all profilers and dispatching at runtime" OFF)
```
Did we discuss that name "single"?  I suppose we can revisit later.

crtrott replied:
not sure we discussed but yeah I agree we can revisit. MONOLITHIC might be another option? Or do it the other way around and name the option KokkosTools_ENABLE_SEPARATE_LIBS or so, which by default is ON?

vlkale replied:
Using KokkosTools_ENABLE_MONOLITHIC makes most sense to me, JM2C.

dalg24 commented on CMakeLists.txt:
```diff
@@ -0,0 +1,220 @@
+mark_as_advanced(KokkosTools_REUSE_KOKKOS_COMPILER)
+
+# Fetch Kokkos options:
+acquire_kokkos_config()
```
When is this useful/desirable?

crtrott replied:
Is this used for stuff like the connector tools to find the same CUDA library? Also there is some tools which have their own "ENABLE_CUDA" etc. e.g. the nvprof thing enabling is based on Kokkos_ENABLE_CUDA, and APEX use of CUPTI is based on the Kokkos option.

vlkale replied:
I think so, I will dig in.

vlkale replied:
The `acquire_kokkos_config()` is used to show/print the kokkos configuration to the user, and to potentially set underlying (e.g.,CUPTI) tools options (e.g., ENABLE_CUDA).

vlkale replied:
I can't find any connector tools right now making use of this from a quick look. I am leaving this line in for now.

dalg24 commented on CMakeLists.txt:
```diff
@@ -0,0 +1,220 @@
+if(KokkosTools_ENABLE_MPI)
+  find_package(MPI REQUIRED)
+  message(STATUS "Found MPI ${MPI_CXX_VERSION}: ${MPI_CXX_LIBRARIES}")
+  set(USE_MPI 1)
+else()
+  set(USE_MPI 0)
```
OK for now but we will need to prefix these

dalg24 commented on CMakeLists.txt:
```diff
@@ -0,0 +1,220 @@
+set(EXPORT_LIB_DIR ${CMAKE_INSTALL_LIBDIR})
+set(EXPORT_TARGETS "" CACHE STRING "" FORCE)
+
+if(WIN32 OR OSX)
```
OSX is not a CMake variable and the message below would be wrong anyway.

vlkale replied:
This is now fixed in NexGenAnalytics:cmake-build-system FYI.

vlkale committed:
Apple Error message and using Apple CMAKE variable

Apple error and using Apple CMAKE variable

fnrizzi committed:
Merge pull request #3 from vlkale/patch-2

Apple Error message and using Apple CMAKE variable

vlkale committed:
Merge branch 'kokkos:master' into cmake-build-system

vlkale committed:
Merge branch 'kokkos:master' into patch-2

vlkale committed:
Merge pull request #1 from vlkale/patch-2

Patch 2

vlkale committed:
More fixes for cleaner builds based on feedback 

More fixes for cleaner builds based on feedback from other Kokkos developers specifically @dalg24

fnrizzi committed:
Merge pull request #4 from vlkale/cmake-build-system

Fixes in CMakeLists.txt based on feedback

vlkale committed:
Error out if Caliper or Apex not found 

Error out if Caliper or Apex not found , fix USE_MPI to KokkosTools_HAS_MPI  (followingthe status message  at the end of file printing value of USE_MPI and the naming convention established.). 

Notes: 
* Exit out of the build if Caliper and Apex are not found. 
* If MPI isn't found, the build just continues without MPI enabled.

vlkale committed:
Putting in Makefile

vlkale committed:
Create Makefile alongside CmakeLists

vlkale committed:
Makefile add to cmakelists.txt

vlkale committed:
Create Makefile

vlkale committed:
Create Makefile

vlkale committed:
Create Makefile

vlkale committed:
Create Makefile

vlkale committed:
Create Makefile

vlkale committed:
HWM Makefile added

vlkale committed:
Create Makefile

vlkale committed:
Create Makefile

vlkale committed:
Create Makefile

vlkale committed:
Delete profiling/vtune-connector/MakefileCXX=icpc CXXFLAGS=-O3 -std=c++11 -g -I$(VTUNE_HOME)/include LDFLAGS=-L$(VTUNE_HOME) directory

vlkale committed:
Create Makefile

vlkale committed:
Create Makefile

vlkale committed:
Create Makefile

vlkale committed:
Create Makefile

vlkale committed:
Put in build instructions

vlkale committed:
Merge pull request #8 from vlkale/vlkale-Patch-cmake-buildsys-with-Makefiles

putting back in  makefile as fallback for CMakelists
